### ISSUE TITLE
BUG: 1. Run the command vak prep gy6or6_train.toml  in the terminal

### ISSUE THREAD

**Issue by cantonsir** [vocalpy/vak#738]
Hi [NickleDave](https://github.com/NickleDave),

It is so wired and the problem is hard to solve.
And i have tried to install the package many time and reintall the conda environment. Could you share the "environment.yml" which could run the code succesfully.

Thanks.
```
(vak-env) C:\Users\szheng2\Documents\Bird_Data\Model Data>vak prep gy6or6_train.toml
Traceback (most recent call last):
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 511, in loads
    ret = decoder.load_line(line, currentlevel, multikey,
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 778, in load_line    value, vtype = self.load_value(pair[1], strictly_valid)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 910, in load_value
    raise ValueError("This float doesn't have a leading "
ValueError: This float doesn't have a leading digit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\config\parse.py", line 169, in _load_toml_from_path
    config_toml = toml.load(fp)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 156, in load
    return loads(f.read(), _dict, decoder)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 514, in loads
    raise TomlDecodeError(str(err), original, pos)
toml.decoder.TomlDecodeError: This float doesn't have a leading digit (line 8 column 1 char 344)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\Scripts\vak.exe\__main__.py", line 7, in <module>
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\__main__.py", line 48, in main
    cli.cli(command=args.command, config_file=args.configfile)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\cli\cli.py", line 49, in cli
    COMMAND_FUNCTION_MAP[command](toml_path=config_file)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\cli\cli.py", line 23, in prep
    prep(toml_path=toml_path)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\cli\prep.py", line 85, in prep
    config_toml = _load_toml_from_path(toml_path)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\config\parse.py", line 171, in _load_toml_from_path
    raise Exception(f"Error when parsing .toml config file: {toml_path}") from e
Exception: Error when parsing .toml config file: gy6or6_train.toml
```

![Fig1](https://github.com/vocalpy/vak/assets/32240006/4d296f20-94cc-4834-a551-a192e67a26b5)

And here is the "conda list"
```
(vak-env) C:\Users\szheng2\Documents\Bird_Data\Model Data>conda list
# packages in environment at C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env:
#
# Name                    Version                   Build  Channel
absl-py                   2.1.0                    pypi_0    pypi
attrs                     23.2.0                   pypi_0    pypi
birdsong-recognition-dataset 0.3.2.post1              pypi_0    pypi
blas                      1.0                         mkl
ca-certificates           2023.12.12           haa95532_0
cachetools                5.3.2                    pypi_0    pypi
certifi                   2024.2.2         py39haa95532_0
cffi                      1.16.0                   pypi_0    pypi
charset-normalizer        2.0.4              pyhd3eb1b0_0
click                     8.1.7                    pypi_0    pypi
cloudpickle               3.0.0                    pypi_0    pypi
colorama                  0.4.6                    pypi_0    pypi
contourpy                 1.2.0                    pypi_0    pypi
crowsetta                 3.4.3                    pypi_0    pypi
cuda-cccl                 12.3.101                      0    nvidia
cuda-cudart               11.8.89                       0    nvidia
cuda-cudart-dev           11.8.89                       0    nvidia
cuda-cupti                11.8.87                       0    nvidia
cuda-libraries            11.8.0                        0    nvidia
cuda-libraries-dev        11.8.0                        0    nvidia
cuda-nvrtc                11.8.89                       0    nvidia
cuda-nvrtc-dev            11.8.89                       0    nvidia
cuda-nvtx                 11.8.86                       0    nvidia
cuda-profiler-api         12.3.101                      0    nvidia
cuda-runtime              11.8.0                        0    nvidia
cycler                    0.12.1                   pypi_0    pypi
dask                      2024.1.1                 pypi_0    pypi
evfuncs                   0.3.5.post1              pypi_0    pypi
filelock                  3.13.1           py39haa95532_0
fonttools                 4.48.1                   pypi_0    pypi
freetype                  2.12.1               ha860e81_0
fsspec                    2024.2.0                 pypi_0    pypi
gmpy2                     2.1.2            py39h7f96b67_0
google-auth               2.27.0                   pypi_0    pypi
google-auth-oauthlib      1.2.0                    pypi_0    pypi
grpcio                    1.60.1                   pypi_0    pypi
idna                      3.4              py39haa95532_0
importlib-metadata        7.0.1                    pypi_0    pypi
importlib-resources       6.1.1                    pypi_0    pypi
intel-openmp              2023.1.0         h59b6b97_46320
jinja2                    3.1.3            py39haa95532_0
joblib                    1.3.2                    pypi_0    pypi
jpeg                      9e                   h2bbff1b_1
kiwisolver                1.4.5                    pypi_0    pypi
lerc                      3.0                  hd77b12b_0
libcublas                 11.11.3.6                     0    nvidia
libcublas-dev             11.11.3.6                     0    nvidia
libcufft                  10.9.0.58                     0    nvidia
libcufft-dev              10.9.0.58                     0    nvidia
libcurand                 10.3.4.107                    0    nvidia
libcurand-dev             10.3.4.107                    0    nvidia
libcusolver               11.4.1.48                     0    nvidia
libcusolver-dev           11.4.1.48                     0    nvidia
libcusparse               11.7.5.86                     0    nvidia
libcusparse-dev           11.7.5.86                     0    nvidia
libdeflate                1.17                 h2bbff1b_1
libjpeg-turbo             2.0.0                h196d8e1_0
libnpp                    11.8.0.86                     0    nvidia
libnpp-dev                11.8.0.86                     0    nvidia
libnvjpeg                 11.9.0.86                     0    nvidia
libnvjpeg-dev             11.9.0.86                     0    nvidia
libpng                    1.6.39               h8cc25b3_0
libtiff                   4.5.1                hd77b12b_0
libuv                     1.44.2               h2bbff1b_0
libwebp-base              1.3.2                h2bbff1b_0
locket                    1.0.0                    pypi_0    pypi
lz4-c                     1.9.4                h2bbff1b_0
markdown                  3.5.2                    pypi_0    pypi
markupsafe                2.1.3            py39h2bbff1b_0
matplotlib                3.8.2                    pypi_0    pypi
mkl                       2023.1.0         h6b88ed4_46358
mkl-service               2.4.0            py39h2bbff1b_1
mkl_fft                   1.3.8            py39h2bbff1b_0
mkl_random                1.2.4            py39h59b6b97_0
mpc                       1.1.0                h7edee0f_1
mpfr                      4.0.2                h62dcd97_1
mpir                      3.0.0                hec2e145_1
mpmath                    1.3.0            py39haa95532_0
networkx                  3.1              py39haa95532_0
numpy                     1.26.3           py39h055cbcc_0
numpy-base                1.26.3           py39h65a83cf_0
oauthlib                  3.2.2                    pypi_0    pypi
openjpeg                  2.4.0                h4fc8c34_0
openssl                   3.0.13               h2bbff1b_0
packaging                 23.2                     pypi_0    pypi
pandas                    2.2.0                    pypi_0    pypi
partd                     1.4.1                    pypi_0    pypi
pillow                    10.2.0           py39h2bbff1b_0
pip                       23.3.1           py39haa95532_0
protobuf                  4.23.4                   pypi_0    pypi
pyasn1                    0.5.1                    pypi_0    pypi
pyasn1-modules            0.3.0                    pypi_0    pypi
pycparser                 2.21                     pypi_0    pypi
pyparsing                 3.1.1                    pypi_0    pypi
python                    3.9.18               h1aa4202_0
python-dateutil           2.8.2                    pypi_0    pypi
pytorch                   2.2.0           py3.9_cuda11.8_cudnn8_0    pytorch
pytorch-cuda              11.8                 h24eeafa_5    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pytz                      2024.1                   pypi_0    pypi
pyyaml                    6.0.1            py39h2bbff1b_0
requests                  2.31.0           py39haa95532_1
requests-oauthlib         1.3.1                    pypi_0    pypi
rsa                       4.9                      pypi_0    pypi
scipy                     1.12.0                   pypi_0    pypi
setuptools                68.2.2           py39haa95532_0
six                       1.16.0                   pypi_0    pypi
soundfile                 0.12.1                   pypi_0    pypi
sqlite                    3.41.2               h2bbff1b_0
sympy                     1.12             py39haa95532_0
tbb                       2021.8.0             h59b6b97_0
tensorboard               2.15.1                   pypi_0    pypi
tensorboard-data-server   0.7.2                    pypi_0    pypi
toml                      0.10.2                   pypi_0    pypi
toolz                     0.12.1                   pypi_0    pypi
torchaudio                2.2.0                    pypi_0    pypi
torchvision               0.17.0                   pypi_0    pypi
tqdm                      4.66.1                   pypi_0    pypi
tweetynet                 0.9.0                    pypi_0    pypi
typing_extensions         4.9.0            py39haa95532_1
tzdata                    2023.4                   pypi_0    pypi
urllib3                   2.1.0            py39haa95532_0
vak                       0.8.2                    pypi_0    pypi
vc                        14.2                 h21ff451_1
vs2015_runtime            14.27.29016          h5e58377_2
werkzeug                  3.0.1                    pypi_0    pypi
wheel                     0.41.2           py39haa95532_0
xz                        5.4.5                h8cc25b3_0
yaml                      0.2.5                he774522_0
zipp                      3.17.0                   pypi_0    pypi
zlib                      1.2.13               h8cc25b3_0
zstd                      1.5.5                hd43e919_0
```

**Comment by NickleDave** [comment]
Hi @cantonsir, sorry you're having this issue.

I don't think this is an issue with your environment.  
The error message makes me think there's a mistake in the TOML config file.  

```
toml.decoder.TomlDecodeError: This float doesn't have a leading digit (line 8 column 1 char 344)
```

If you are using the config from the tutorial, line 8 would be
```
data_dir = "/PATH/TO/FOLDER/gyor6/032212"
```

Maybe you changed it but you did not put quotes around the string, or something like that?

Can you please reply with the content of your config file?  
Either by pasting it directly into a comment or by attaching it to a comment.  
That way I can help you troubleshoot.

**Comment by cantonsir** [comment]
Hi [NickleDave](https://github.com/NickleDave),

The following are part of "gy6or6_train.toml".

Thanks for your help.

```
# PREP: options for preparing dataset
[PREP]
# dataset_type: corresponds to the model family such as "frame classification" or "parametric umap"
dataset_type = "frame classification"
# input_type: input to model, either audio ("audio") or spectrogram ("spect")
input_type = "spect"
# data_dir: directory with data to use when preparing dataset
data_dir = r"C:\Users\szheng2\Documents\Bird_Data\Model Data\gy6or6\032212"
# output_dir: directory where dataset will be created (as a sub-directory within output_dir)
output_dir = r"C:\Users\szheng2\Documents\Bird_Data\Model Data\vak\prep\train"
# audio_format: format of audio, either wav or cbin
audio_format = "wav"
# annot_format: format of annotations
annot_format = "simple-seq"
# labelset: string or array with unique set of labels used in annotations
labelset = "iabcdefghjk"
# train_dur: duration of training split in dataset, in seconds
train_dur = 50
# val_dur: duration of validation split in dataset, in seconds
val_dur = 15
# test_dur: duration of test split in dataset, in seconds
test_dur = 30

# SPECT_PARAMS: parameters for computing spectrograms
[SPECT_PARAMS]
# fft_size: size of window used for Fast Fourier Transform, in number of samples
fft_size = 512
# step_size: size of step to take when computing spectra with FFT for spectrogram
# also known as hop size
step_size = 64
```

**Comment by NickleDave** [comment]
Ah I think I see.

You want to remove the `r` from in front of all the options that are strings.

As you know the prefix `r` creates raw strings in Python. I think you are doing this because you have had issues with Windows paths in Python before.

But the config file is not Python code; it's in [TOML](https://toml.io/en/) format. There's no such thing as a "raw" string in TOML.

In addition, we use [pathlib](https://docs.python.org/3/library/pathlib.html) to work with paths. That is once `vak` parses the TOML config file and loads paths like `data_path` into Python, they are `pathlib.Path` instances. pathlib lets you write paths with single forward slashes on Windows, and does all the work for you so that the paths are parsed correctly. E.g. see https://www.reddit.com/r/learnpython/comments/sifzlv/windows_path_to_python_path/

In other words, you should just be able to remove the `r` and it will work. Please let me know if not.

**Comment by cantonsir** [comment]
Hi

I have tried it before, and i try it again, and it still doesn't work.

```
# PREP: options for preparing dataset
[PREP]
# dataset_type: corresponds to the model family such as "frame classification" or "parametric umap"
dataset_type = "frame classification"
# input_type: input to model, either audio ("audio") or spectrogram ("spect")
input_type = "spect"
# data_dir: directory with data to use when preparing dataset
data_dir = "C:/Users/szheng2/Documents/Bird_Data/Model Data/gy6or6/032212"
# output_dir: directory where dataset will be created (as a sub-directory within output_dir)
output_dir = "C:/Users/szheng2/Documents/Bird_Data/Model Data/vak/prep/train"
# audio_format: format of audio, either wav or cbin
audio_format = "wav"
# annot_format: format of annotations
annot_format = "simple-seq"
# labelset: string or array with unique set of labels used in annotations
labelset = "iabcdefghjk"
# train_dur: duration of training split in dataset, in seconds
train_dur = 50
# val_dur: duration of validation split in dataset, in seconds
val_dur = 15
# test_dur: duration of test split in dataset, in seconds
test_dur = 30
```

```
(vak-env) C:\Users\szheng2\Documents\Bird_Data\Model Data>vak prep gy6or6_train.toml
Traceback (most recent call last):
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 511, in loads
    ret = decoder.load_line(line, currentlevel, multikey,
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 778, in load_line    value, vtype = self.load_value(pair[1], strictly_valid)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 910, in load_value
    raise ValueError("This float doesn't have a leading "
ValueError: This float doesn't have a leading digit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\config\parse.py", line 169, in _load_toml_from_path
    config_toml = toml.load(fp)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 156, in load
    return loads(f.read(), _dict, decoder)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\toml\decoder.py", line 514, in loads
    raise TomlDecodeError(str(err), original, pos)
toml.decoder.TomlDecodeError: This float doesn't have a leading digit (line 37 column 1 char 1635)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\Scripts\vak-script.py", line 9, in <module>
    sys.exit(main())
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\__main__.py", line 48, in main
    cli.cli(command=args.command, config_file=args.configfile)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\cli\cli.py", line 54, in cli
    COMMAND_FUNCTION_MAP[command](toml_path=config_file)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\cli\cli.py", line 28, in prep
    prep(toml_path=toml_path)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\cli\prep.py", line 85, in prep
    config_toml = _load_toml_from_path(toml_path)
  File "C:\Users\szheng2\AppData\Local\anaconda3\envs\vak-env\lib\site-packages\vak\config\parse.py", line 171, in _load_toml_from_path
    raise Exception(
Exception: Error when parsing .toml config file: gy6or6_train.toml
```

**Comment by cantonsir** [comment]
Finally, it is work.

I need to check all the path, including PATH from the [Train] part.

Thanks a lot.

**Comment by NickleDave** [comment]
Great, glad to hear it @cantonsir.

That was my guess from the new error message from where it said "line 37".  

I know error messages can be overwhelming when they dump out a ton of text, but I have learned the hard way it's worth the effort to read them :slightly_smiling_face: even if sometimes the programmer doesn't do a good job of making them very helpful :innocent: 

Please just let us know if there's anything we can do to help you.

You can join our forum here, introduce yourself, and ask more general question too if you'd like: https://forum.vocalpy.org/
